Tests template

hussainmansour opened the pull request.

josephShokry committed:
parsing input using cmd

adding example of how to parse input using cmd

josephShokry committed:
handling json

adding example of handling json data

josephShokry committed:
adding the main file

adding the main file parsing the input (the command and the json) and check the command then execute the proper function

josephShokry committed:
adding more input variables

josephShokry committed:
adding the create function

- adding the create function that makes the directories using the schema
- modifying how to get the schema instead of getting it from the cmd directly now we get it through a text file

josephShokry committed:
create funciton

completing the create function, making the file of the parser and cleaning the code

josephShokry committed:
handling errors of invalid schema file

if the user parsed a not existent file of not parsing the schema file at all

josephShokry committed:
Delete __pycache__ directory

josephShokry committed:
adding command factory and help in parser

i merged the code of hussain about the command factory with the parser the main code and adder the help attribute of the parsers

josephShokry committed:
.

josephShokry committed:
Merge branch 'tests-template' of https://github.com/josephShokry/SimpleFSDB into tests-template

josephShokry committed:
adding the keys class

adding the keys class to simplify the coding

josephShokry committed:
adding test cases

adding test cases of create function to check if all the tables and DB has been created or not

hussainmansour committed:
merging between test-template and AvoidingError

josephShokry committed:
Delete __pycache__ directory

josephShokry committed:
Delete tests directory

bangkook commented on commands_function.py:
```diff
@@ -0,0 +1,59 @@
+class commandfactory:
+    @staticmethod
+    def build_command(args):
+        command_type=args.command
```
add space before and after the '=' operator

josephShokry replied:
ok we will

mohamedabdelsalam replied:
You are create command not build it. Because there is two design battern regarding to object creation :- 
1- Factory design pattern 
2- Builder design pattern. 

Here, you are creating the object, not building it

josephShokry replied:
@mohamedabdelsalam sorry can you explain more.

bangkook commented on commands_function.py:
```diff
@@ -0,0 +1,59 @@
+        if not os.path.isdir(path): #to check if the DB directory is created before or not
+            os.mkdir(path)
+        for table in schima[keys.tables]:
+            table_path=os.path.join(path,table[keys.name])
```
same previous comment for all assignments

josephShokry replied:
ok!

bangkook commented on commands_function.py:
```diff
@@ -0,0 +1,59 @@
+        self.schema=schema
+    def excute(self):
+        schema=self.schema
+        f=open(schema,"r")
```
don't name a variable 'f' only

josephShokry replied:
thanks for the mention

mohamedabdelsalam replied:
+1

bangkook commented on main.py:
```diff
@@ -0,0 +1,9 @@
+
+if __name__=='__main__':
+    args=Parser.parseInput()
+    command =commands_function.commandfactory.build_command(args)
```
you can use
"from commands_function import commandfactory"
to call commandfactory directly without referring to the file name

bangkook replied:
same with Parser

josephShokry replied:
this is a good one, I didn't know about it before.
I will.

mohamedabdelsalam commented on Parser.py:
```diff
@@ -0,0 +1,18 @@
+import os
+
+parser = argparse.ArgumentParser()
+parent_path=os.getcwd()
```
Why do we need this ?

josephShokry replied:
sorry I forgot to delete them

mohamedabdelsalam commented on Parser.py:
```diff
@@ -0,0 +1,18 @@
+parent_path=os.getcwd()
+def parseInput():
+    parser=argparse.ArgumentParser()
+    parser.add_argument("-cmd","--command",type=str.lower,help="the quiry command",required=True,choices=["get","set","create","delete"])
```
follow convension, space after `,`

josephShokry replied:
done!

mohamedabdelsalam commented on Parser.py:
```diff
@@ -0,0 +1,18 @@
+def parseInput():
+    parser=argparse.ArgumentParser()
+    parser.add_argument("-cmd","--command",type=str.lower,help="the quiry command",required=True,choices=["get","set","create","delete"])
+    parser.add_argument("-sch","--schema",type=str,required=(parser.parse_known_args()[0].command=="create"),help="path to your schema file")
```
space after `=` and before in all lines

josephShokry replied:
did it.

mohamedabdelsalam commented on commands_function.py:
```diff
@@ -0,0 +1,59 @@
+    def excute():
+        """the interface Icommand"""
+
+class commandfactory:
```
file name should be `command_factory` not `commands_factory`

josephShokry replied:
ok done

mohamedabdelsalam commented on commands_function.py:
```diff
@@ -0,0 +1,59 @@
+class commandfactory:
+    @staticmethod
+    def build_command(args):
+        command_type=args.command
```
lower case to the command_type

mohamedabdelsalam commented on commands_function.py:
```diff
@@ -0,0 +1,59 @@
+        elif command_type == "set":
+            return set()
+        
+class create(Icommand):
```
name it as `CreateCommand` not a Create. + class name should start with Capital

mohamedabdelsalam commented on commands_function.py:
```diff
@@ -0,0 +1,59 @@
+            return set()
+        
+class create(Icommand):
+    def __init__(self,schema):
```
space after `,`

mohamedabdelsalam commented on commands_function.py:
```diff
@@ -0,0 +1,59 @@
+        
+class create(Icommand):
+    def __init__(self,schema):
+        self.schema=schema
```
space after `=`

mohamedabdelsalam replied:
Is this schema or schema_file ? it is confusing name

mohamedabdelsalam commented on commands_function.py:
```diff
@@ -0,0 +1,59 @@
+    def excute(self):
+        schema=self.schema
+        f=open(schema,"r")
+        schima=json.load(f)
```
schema not schima 

What happend if the file doesn't exists ? 

I would recommend create a private method here, this method takes a schema_file and return schema object

mohamedabdelsalam commented on commands_function.py:
```diff
@@ -0,0 +1,59 @@
+        parent_dir=os.getcwd() #get the current dir of the project
+        data_base_name=schima[keys.DB_name]
+        path=os.path.join(parent_dir,data_base_name)
+        if not os.path.isdir(path): #to check if the DB directory is created before or not
+            os.mkdir(path)
```
again, why don't u use mkdir with boolean flag, you don't need to use isDir

josephShokry replied:
ok, but I don't know what is wrong with the isDir function.

mohamedabdelsalam commented on keys.py:
```diff
@@ -0,0 +1,10 @@
+class keys:
```
class name is confusing. 
fllow the convension as well 

- constants in the class should be capital, e.g SCHEMA, TABLES, etc 

- Also search how to make sure in python they are constant

josephShokry replied:
ok did it

mohamedabdelsalam commented on schema.txt:
```diff
@@ -0,0 +1,19 @@
+{
```
Why you didn't create a folder call tests and move it there ?

josephShokry replied:
I did but did not push it to GitHub
and did not put this file in it because it is an essential file 
but after mentioning it and thinking again you are right and I will.

mohamedabdelsalam commented on commands_function.py:
```diff
@@ -0,0 +1,59 @@
+import os
```
Create file called ICommand 
file called command_factory
and file for each command

mohamedabdelsalam commented on test_cases.py:
```diff
@@ -0,0 +1,22 @@
+
+parent_path=os.getcwd()
+
+class test_functions(unittest.TestCase):
```
You are not handling al test cases, please cover more cases and more edge cases

josephShokry committed:
handling some comments

mohamedabdelsalam commented on Parser.py:
```diff
@@ -0,0 +1,17 @@
+    
+
+    return parser.parse_args()  
+    
+ 
```
extra vertical spaces

mohamedabdelsalam commented on Parser.py:
```diff
@@ -0,0 +1,17 @@
+    parser.add_argument("-val", "--value", type = str, help = "the value that you want to change or add to the DB")  
+    parser.add_argument("-db", "--data_base", type = str, help = "the database name")
+    parser.add_argument("-tb", "--table", type = str, help = "the table name that you want to access")  
+    
+
```
extra vertical spaces

mohamedabdelsalam commented on command_function.py:
```diff
@@ -0,0 +1,59 @@
+    def build_command(args):
+        command_type=args.command
+        if command_type == "create":
+            return create(args.schema)
```
Is this schema or schema_file ?

mohamedabdelsalam commented on command_function.py:
```diff
@@ -0,0 +1,59 @@
+        elif command_type == "set":
+            return set()
+        
+class create(Icommand):
```
- class name should start with Capital letter 
- make every class in a file
- put al commands classes in one folder

EhabYasser25 replied:
Ok, I will handle this.

mohamedabdelsalam commented on command_function.py:
```diff
@@ -0,0 +1,59 @@
+    def __init__(self, schema):
+        self.schema=schema
+    def excute(self):
+        schema = self.schema
+        schema_file = open(schema, "r")
+        schema = json.load(schema_file)
```
move this in constucrtor

mohamedabdelsalam commented on command_function.py:
```diff
@@ -0,0 +1,59 @@
+        parent_dir = os.getcwd() #get the current dir of the project
+        data_base_name = schema[keys.DB_NAME]
+        path = os.path.join(parent_dir, data_base_name)
+        if not os.path.isdir(path): #to check if the DB directory is created before or not
```
use mkdir(path, true)

EhabYasser25 replied:
I dont understand exactly, do I need to use mkdir(path, true) instead of os.mkdir(path)?

mohamedabdelsalam commented on command_function.py:
```diff
@@ -0,0 +1,59 @@
+            table_path = os.path.join(path, table[keys.NAME])
+            if not os.path.isdir(table_path):
+                os.mkdir(table_path)
+                # file_name=os.path.join(table_path, "info.txt")
```
why do we have this comments ?

josephShokry replied:
this is the code used to make table schema details (the names of columns, the name of the primary key, . . )
I am not sure if we really need it in the future and in what format I will need this file so I made this comment in case I need it in the future I can remove the comment on these lines and edit it.

mohamedabdelsalam commented on command_function.py:
```diff
@@ -0,0 +1,59 @@
+            os.mkdir(path)
+        for table in schema[keys.TABLES]:
+            table_path = os.path.join(path, table[keys.NAME])
+            if not os.path.isdir(table_path):
```
use mkdir(path, true)

mohamedabdelsalam commented on schema_keys.py:
```diff
@@ -0,0 +1,24 @@
+from tkinter.tix import COLUMN
+
+
+class keys:
```
class name capital

mohamedabdelsalam commented on schema_keys.py:
```diff
@@ -0,0 +1,24 @@
+from imp import PKG_DIRECTORY
+from msilib import schema
+from tkinter.tix import COLUMN
+
```
Why do we need those imports ?

mohamedabdelsalam commented on Parser.py:
```diff
@@ -0,0 +1,17 @@
+import argparse
+import os
+
+
```
extra vertical spaces

josephShokry committed:
addressing some comments

Commit:
Handling All comments on PR

mohamedabdelsalam commented on CommandsFunctions/CommandFactory.py:
```diff
@@ -0,0 +1,17 @@
+
+class CommandFactory:
+    @staticmethod
+    def build_command(args):
```
it is not build, it create

mohamedabdelsalam commented on CommandsFunctions/CommandFactory.py:
```diff
@@ -0,0 +1,17 @@
+    def build_command(args):
+        command_type=args.command
+        if command_type == "create":
+            return Create(args.schema_file)
```
`CreateCommand` not `Create`. class name can't be a verb

mohamedabdelsalam commented on CommandsFunctions/Create.py:
```diff
@@ -0,0 +1,36 @@
+from Icommand import Icommand
+
+class Create(Icommand):
+    def __init__(self, schema_file):
```
too much try and catch isn't good. 
it is better to have validate function. It throws exception in case anything error, then execute your code

mohamedabdelsalam replied:
look at @bangkook code as an example. I did a comment in this part there

mohamedabdelsalam commented on CommandsFunctions/Create.py:
```diff
@@ -0,0 +1,36 @@
+        parent_dir = os.getcwd() #get the current dir of the project
+        data_base_name = schema_data[Keys.DB_NAME]
+        path = os.path.join(parent_dir, data_base_name)
+        if not os.path.isdir(path): #to check if the DB directory is created before or not
+            os.mkdir(path, True)
```
you use True flag, why do you need to check ?

mohamedabdelsalam commented on CommandsFunctions/Create.py:
```diff
@@ -0,0 +1,36 @@
+                table_path = os.path.join(path, table[Keys.NAME])
+                if not os.path.isdir(table_path):
+                    os.mkdir(table_path, True)
+                    # file_name = os.path.join(table_path, "info.txt")
+                    # f = open(file_name, 'w')
+                    # f.write(json.dumps(i,  indent = 4))
```
remove commnets

mohamedabdelsalam commented on CommandsFunctions/Create.py:
```diff
@@ -0,0 +1,36 @@
+        try:
+            for table in schema_data[Keys.TABLES]:
+                table_path = os.path.join(path, table[Keys.NAME])
+                if not os.path.isdir(table_path):
```
no need for if condition here

mohamedabdelsalam commented on CommandsFunctions/Create.py:
```diff
@@ -0,0 +1,36 @@
+                    # f = open(file_name, 'w')
+                    # f.write(json.dumps(i,  indent = 4))
+        except:
+            print('No Tables Found')
```
Don't print inside the class

mohamedabdelsalam commented on CommandsFunctions/Get.py:
```diff
@@ -0,0 +1,8 @@
+
+class Get(Icommand):
+    def execute(self):
+        print("dummy get will be implemented later")
```
raise error

mohamedabdelsalam commented on SchemaKeys.py:
```diff
@@ -0,0 +1,13 @@
+from msilib import schema
+from tkinter.tix import COLUMN
+
+class Keys:
```
This class should be abstract

mohamedabdelsalam commented on TestCases.py:
```diff
@@ -0,0 +1,22 @@
+import unittest
```
write more unit test to test failure senatios

josephShokry committed:
adding more test cases for the create function

adding a folder "testCases_schemas" to add the test schemas in it .
adding more test cases for the create functoin.

josephShokry committed:
Create .gitignore
